### PR TITLE
[24201] Add permission edit_own_work_packages

### PR DESCRIPTION
**Pull request by silkentrance**
Add permission edit_own_work_packages so that users can edit their reported and subsequently also owned work packages.

https://community.openproject.com/work_packages/24201

Todos
- [ ] Range filter in one row instead of two

### CODE REVIEW

**Comment by silkentrance** [comment]
The PR has been provided to you for review and you may have some additional comments or advise for me on how to make it better a/o implement the test cases for it. Please note that it is still missing required test cases.

**Comment by ulferts** [comment]
@silkentrance I saw the PR but I'll need some time to wrap my head around the problem. Will get back to you.

**Comment by silkentrance** [comment]
@ulferts thank you. I am currently working on the failing tests caused by the change in behaviour of the available_projects_on_edit... API call. While the frontend behaves just fine, the existing tests will fail.

**Comment by silkentrance** [comment]
@ulferts the use case is rather simple. considering the official community.openproject.org site where I can add new work packages to the public projects, I cannot edit these work packages after they have been created, i.e. where the user's role misses the edit_work_packages permission, it will limit the user to just the work packages that he or she is the author of, provided that the role is granted the edit_own_work_packages permission.

This fixes a number of issues, most prevalently the issue where I can quick edit/add a work package from the work package list, enter its description and then it will already be persisted to the database. After which I can no longer edit the rest of the data. Not to mention the ability to edit/revise the summary/description/type/version etc. after a work package has been created.

**Comment by silkentrance** [comment]
@ulferts the change to available_projects_on_edit is meant to only provide the user with a resonable default, otherwise the drop down will be empty, in case that the user's role does not include the move_work_packages permission.

Speaking of which. I have recently switched to RubyMine for development purposes. However, I am currently not able to debug the code blocks in the tests that contain the actual expectations, e.g.

```
1  context 'w/ the edit_own_work_packages permission' do
2    it {
3      expect(response.status).to eq(200)
4    }
5
6    it 'has the project for which the edit_own_work_packages permission exists' do
7      expect(response.body).to be_json_eql(owned_wp_project.id).at_path('_embedded/elements//id')
8    end
9  end
```

image

Setting a breakpoint on either line 3 or 7 will not allow me to debug. Setting it on either 2 or 6 will cause the debugger to step in. However, stepping into or skipping to the next breakpoint will yield no results.

This happens during both remote and local debug.

I think that this has to do something with the actual expectations being run in a different process, however, I am quite new to ruby development.

What IDE are you using?

P.S. yes, I have looked on google but found no solution to the problem.

**Comment by silkentrance** [comment]
@ulferts Tests have been added. I am currently trying to flesh out newly introduced code climate issues.

End-to-end tests are still missing, but I think that by testing the available_projects_on_edit service api should suffice, don't you?

**Comment by silkentrance** [comment]
@ulferts the user authorization service tends to become a hotspot with a currently rather high complexity and, given additional permissions added to the system, also of increasing complexity. This, together with the current approach of having to know about the required permissions in a specific use case a/o usage scenario, e.g. 

```
    if EXPORT_FORMATS.include?(params[:format]) &&
       !User.current.allowed_to?(:export_work_packages, @project, global: @project.nil?)
```

or even more complex

```
        resource :available_projects do
          before do
            begin
              authorize(:edit_work_packages, context: @work_package.project)
            rescue API::Errors::Unauthorized
              authorize(:edit_own_work_packages, context: @work_package)
            end
          end
```

all add to the complexity of the application instead of making it simpler, making future patches by for example plugins rather difficult.

I see that you and your team have begun already refactoring the whole authorization part. It is, however, quite difficult to understand, especially since that there are many delegations in the code, which of course all end up either in Authorization or the UserAllowedService. Perhaps I can contribute some additional ideas, if you like.

**Comment by ulferts** [comment]
Took me some time @silkentrance but have made up my mind over the weekend. This will be long, please bear with me.

## Background

The current authorization system code is implemented with the following priorities in mind (most important to less important): 
 * performance
 * extensibility with regards to additional permissions
 * extensibility with regards to plugins
 * readability/maintainability

Performance earns its utmost importance by the sheer number of calls (explicit and implicit) made to the authorization code. On a call to `/api/v3/work_packages` with 50 work packages per page we will have about 2400 calls checking permissions. Because of this, the UserAllowedService always fetches all permissions for any given context from the db and will then calculate the answer from the memorized information. A context is either a specific project or having no project at all. As long as the context remains the same (and as long as the same instance of the user is queried), no further db query is necessary. The service itself is thus a pretty slim layer on top of the actual authorization code.

The bulk of the authorization is carried out in the database. If you take a look at the [Authorization class](https://github.com/opf/openproject/blob/dev/app/services/authorization.rb) you will notice four queries which are executed for different types of authorization requests:
* `Authorization::UserAllowedQuery` gets all users having a certain permission in a project
* `Authorization::ProjectQuery` gets all projects in which a certain user has a permission
* `Authorization::UserProjectRolesQuery` and `Authorization::UserGlobalRolesQuery` get all the roles (and with it the permissions) a user has in a context.

The last two are used in the `UserAllowedService`.

Those query classes generate Arel ASTs which are compatible with rails' db query mechanisms as rails uses Arel under the hood. That way, the generated ASTs can be employed in a bunch of ways, most importantly to serve as scopes (e.g. `visible`) which limit a result set down to not leak information.

Having 4 queries to do the job is undesirable with regards to maintainability. While I believe it to be possible to have only one query to do the job, it will most certainly be slower or one will have to have a lot of logic altering the sql statement which will defeat the purpose of increasing maintainability.

The current implementation is very extensible both by additional permissions (it actually does not care for the permission which is only a string stored in the db) and by plugins which can alter the queries. 

Granted, the implementation is not very readable and by that not very maintainable as it has a couple of layers and requires the developer to understand the principles of the implementation before one is able to make sense of the code. I have written the stuff and I struggle to keep all of it in my head. But I am convinced, that the current implementation is the optimal solution for the time being. 

Please also see https://github.com/opf/openproject/pull/4778 and https://github.com/opf/openproject/pull/4803 as those two PRs introduced the current implementation.

## Remarks on the PR

Because of the high complexity inside the inner workings of the authorization code we should strive to improve the abstraction of that code. And you are right, it is highly unfortunate, that a developer is required to know which permission to query for in a given scenario.

Just to get the definitions straight, in my book you are not simply adding a permission in the current PR. You are actually adding a new context, the work package, as well on which the permission is checked.   

Following that definition, we have to make sure, that subsequent requirements do not force us to add more and more contexts (like Changeset, News, Message) to the UserAllowedService. Instead, it has to be the responsibility of each model to respond to a set of methods charged with enforcing authorization. That set is to be the same for every model. A rough sketch of methods:
* `Model.visible` and `Model.new.visible?`
* `Model.editable` and `Model.new.editable?` (updatable or patchable would be alternatives)
* `Model.deletable` and `Model.new.deletable?` 
* `Model.new.creatable?`

It is important to have both the scope method as well as the method on the individual instance so that one can retrieve a set of instances on the one hand, while also being able to check the permission on an instance.

The currently implemented change in this PR is limited to checking the permission on an individual work package. It will therefore be costly to get all work packages a user can edit. While this might be ok for editing, it is not for the visibility check. The notion of private work packages, where the author is able to see and edit work packages albeit lacking the `view_work_packages` and `edit_work_packages` permission has been a long desired feature. With the feature in place the scenarios in which OpenProject solves the business need (think help desk) will increase dramatically. Only when the mechanism includes a scope however, will it be successful.

The scopes will most likely be project based (`Authorization::ProjectQuery`) while the instance methods are user based (`Authorization::UserProjectRolesQuery` and `Authorization::UserGlobalRolesQuery` via the `UserAllowedService`). 

Because most models already have a project association, using those scopes should be easy. And some models already have a subset of the scopes/methods, e.g. `WorkPackage` with its `visible` scope.

But putting the code inside the models will further bloat the models. We should rather rededicate the policy classes in app/policies to serve as authorization providers. Please see https://github.com/elabs/pundit for more details as the idea was inspired from there. We could also try to employ pundit for our purposes. The code currently found in the policy files should be rewritten to the changed responsibility of the classes. 

For the edit_own_work_packages permission the work package policy class could then look like this:

```
class WorkPackagePolicy < ApplicationPolicy
  class Scopes
    def editable
      all = joins(:project)
              .merge(Project.allowed_to(user, :edit_work_packages))
      own = joins(:project
                 .merge(Project.allowed_to(user, :edit_own_work_packages))
                 .where(author_id: user.id)

      all.or(own)
    end
  end

  def editable?
    User.allowed_to(:edit_work_packages, work_package.project) || 
      (User.allowed_to(:edit_own_work_packages, work_package.project) &&
        work_package.author_id == user.id)
  end
end
```

That way, we will have all the authorization code in a single file for every model and a developer will no longer have to remember the specifics.   

If we build on the policy file idea and include methods simply for permission checking your example could look like this:

```
        resource :available_projects do
          before do
            begin
              @work_package.editable?! 
            end
          end
```

where `editable?!` raises an exception. But as the exception might be context dependent we could also change the `authorize` method raise an exception based on a boolean

```
        resource :available_projects do
          before do
            begin
              authorize @work_package.editable? 
            end
          end
```

Putting the permissions in the policy files will then also enable us to remove this weird :controller => [:action] syntax in the permissions initializer as we can list the permissions eligible for an action inside the policy.  

## tl;dr

I propose putting the authorization scope that is not scoped directly to a user or a project in policy classes that have a fixed set of authorization methods. The methods consist of both scope as well as instance methods. A developer will no longer have to memorize the symbols but can rather simply query the model. The scopes generating DB queries and which are powering all of this can remain the same but will be further hidden behind another layer of abstraction. 

## Open

* Given the authorization interfaced being changed to what is proposed here, how would one check for non model based permissions? - might want to introduce artificial models just for the scope then.


PS: I am using MacVim so I do not benefit from the full power of an IDE.

**Comment by silkentrance** [comment]
@ulferts This is a lot of information. I will have to digest that first, so please be patient with me.

**Comment by WebsiteDeveloper** [comment]
@silkentrance @ulferts any news on this?

**Comment by silkentrance** [comment]
@WebsiteDeveloper none ATM. I will try to get back to this in mid to late february.

**Comment by silkentrance** [comment]
@lindenthal Niels, I no longer have time to pursue this one. Perhaps @ulferts might want to chime in and complete it, especially with the refactoring going on from moving policies to app/policies.

**Comment by lindenthal** [comment]
@ulferts should we keep this PR or delete it?

**Comment by ulferts** [comment]
Removing as stale.
